### PR TITLE
Castorini smmodel

### PR DESCRIPTION
**Pull request by gauravbaruah**
Code bridge between Castor/sm-model and Anserini.

### CODE REVIEW

File: sm-model/SMCastorini.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+from torch.autograd import Variable
+import numpy as np
+import os
+from nltk.tokenize import TreebankWordTokenizer
+from nltk.stem.porter import *
+from collections import defaultdict
+import pickle
+import string
+
+from model import QAModel
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: Import ordering: standard imports, third party imports, local imports
https://www.python.org/dev/peps/pep-0008/#imports

File: sm-model/SMCastorini.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
+import torch
+from torch.autograd import Variable
+import numpy as np
+import os
+from nltk.tokenize import TreebankWordTokenizer
+from nltk.stem.porter import *
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: avoid wildcard imports

File: sm-model/SMCastorini.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        for i in xrange(len(terms)):
+            word = terms[i]        
+            if word not in self.vocab_dict:
+                emb = torch.from_numpy(self.unk_term_vec)
+            else:
+                emb = torch.from_numpy(self.W[self.vocab_dict[word]])                        
+            word_embeddings[i] = emb            
+
+        input_tensor = torch.zeros(1, self.vec_dim, len(terms))
+        input_tensor[0] = torch.transpose(word_embeddings, 0 , 1)
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: remove space

File: sm-model/SMCastorini.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        return input_tensor
+
+
+    def get_tensorized_inputs(self, batch_ques, batch_sents, batch_ext_feats):
+        
+        assert(1 == len(batch_ques) )
+        
+        tensorized_inputs = []
+        for i in xrange(len(batch_ques)):
+            xq = Variable(self.make_input_matrix(batch_ques[i]) )
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: remove extra space before ) line 121, 122

File: sm-model/SMCastorini.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            word_embeddings[i] = emb            
+
+        input_tensor = torch.zeros(1, self.vec_dim, len(terms))
+        input_tensor[0] = torch.transpose(word_embeddings, 0 , 1)
+        return input_tensor
+
+
+    def get_tensorized_inputs(self, batch_ques, batch_sents, batch_ext_feats):
+        
+        assert(1 == len(batch_ques) )
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: extra space before )

File: sm-model/SMCastorini.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            pred = torch.exp(pred)
+            scores_sentences.append( (pred.data.squeeze()[1], a_list[i]) )
+        
+        return scores_sentences
+
+
+if __name__ == "__main__":
+    
+
+    smmodel = SMModelCastorini('../../data/TrecQA/sm.model.aquaint.castorini',
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: delete extra newlines?

File: sm-model/SMCastorini.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                    '../../data/word2vec-models/aquaint+wiki.txt.gz.ndim=50.cache',
+                    'stopwords.txt',
+                    'word2dfs.p')
+    
+    
+    q = "who is the author of the book , `` the iron lady : a biography of margaret thatcher '' ?"
+    a = [ 
+    "the iron lady ; a biography of margaret thatcher by hugo young -lrb- farrar , straus & giroux -rrb-",
+    "in this same revisionist mold , hugo young , the distinguished british journalist , has performed a brilliant dissection of the notion of thatcher as a conservative icon .",
+    "in `` the iron lady , '' young traces the winding staircase of fortune that transformed the younger daughter of a provincial english grocer into the greatest woman political leader since catherine the great .",
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: divide the long lines into multiple lines with `\`?

File: sm-model/SMCastorini.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    "`` he is the very essence of the classless meritocrat , '' says hugo young , thatcher 's biographer .",
+    "from her father , young argues , she inherited a `` joyless earnestness '' that combined with her early interest in science to produce the roots of her public character .",
+    "this is not the answer",
+    "asdfawe asdf sertse dgfsgsfg"
+    ]
+	
+    ss = smmodel.rerank_candidate_answers(q, a)
+    print 'Question:', q
+    for score, sentence in ss:
+        print score, '\t', sentence
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: Are we not using python3? print() instead of print?

File: sm-model/SMCastorini.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+        return tensorized_inputs
+
+
+    def rerank_candidate_answers(self, question, answers):
+        # tokenize
+        q_str, a_list = self.parser(question, answers)
+
+        # calculate overlap features
+        overlap_feats = self.compute_overlap_features(q_str,a_list, stoplist=None, word2df=self.word2dfs)
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: space after `q_str,`

File: sm-model/SMCastorini.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        self.vec_dim = self._preload_cached_embeddings(word_embeddings_cache_file)
+        self.unk_term_vec = np.random.uniform(-0.25, 0.25, self.vec_dim) 
+
+        # stopwords
+        self.stoplist = set([line.strip() for line in open(stopwords_file)])
+
+        # word dfs
+        if os.path.isfile(word2dfs_file):
+		    self.word2dfs = pickle.load( open( "word2dfs.p", "rb" ) )
+
```
Comment (rosequ) [inline sm-model/SMCastorini.py]: Wouldn't you prefer `with` while opening the files?

File: sm-model/SMCastorini.py (hunk 4 above)
Comment (rosequ) [inline sm-model/SMCastorini.py]: if porting to python3, change `xrange` to `range`

File: sm-model/SMCastorini.py (hunk 9 above)
Comment (gauravbaruah) [inline sm-model/SMCastorini.py]: should there not be a space after q_str?

**Pull request merged**